README.md
# Fluentd Cloudwatch Plugin
[![Circle CI](https://circleci.com/gh/sampointer/fluent-plugin-cloudwatch-ingest.svg?style=shield)](https://circleci.com/gh/sampointer/fluent-plugin-cloudwatch-ingest) [![Gem Version](https://badge.fury.io/rb/fluent-plugin-cloudwatch-ingest.svg)](https://badge.fury.io/rb/fluent-plugin-cloudwatch-ingest) ![](http://ruby-gem-downloads-badge.herokuapp.com/fluent-plugin-cloudwatch-ingest?type=total) [![Join the chat at https://gitter.im/fluent-plugin-cloudwatch-ingest](https://badges.gitter.im/fluent-plugin-cloudwatch-ingest.svg)](https://gitter.im/fluent-plugin-cloudwatch-ingest/Lobby?utm_source=share-link&utm_medium=link&utm_campaign=share-link) [![Dependency Status](https://gemnasium.com/badges/github.com/sampointer/fluent-plugin-cloudwatch-ingest.svg)](https://gemnasium.com/github.com/sampointer/fluent-plugin-cloudwatch-ingest)

## Introduction

This gem was created out of frustration with existing solutions for Cloudwatch log ingestion into a Fluentd pipeline. Specifically, it has been designed to support:

* The 0.14.x fluentd plugin API
* Native IAM including cross-account authentication via STS
* Tidy state serialization
* HA configurations without ingestion duplication

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'fluent-plugin-cloudwatch-ingest'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install fluent-plugin-cloudwatch-ingest

## Usage
```
<source>
  @type cloudwatch_ingest
  region us-east-1
  sts_enabled true
  sts_arn arn:aws:iam::123456789012:role/role_in_another_account
  sts_session_name fluentd-dev
  aws_logging_enabled true
  log_group_name_prefix /aws/lambda
  log_stream_name_prefix 2017
  log_group_exclude_regexp [^A-Z]{1}.* # exclude log groups that start with a captial
  state_file_name /mnt/nfs/cloudwatch.state
  interval 60
  max_log_streams_per_group 50
  error_interval 5          # Time to wait between error conditions before retry
  limit_events 10000        # Number of events to fetch in any given iteration
  event_start_time 0        # Do not fetch events before this time (UNIX epoch, miliseconds)
  oldest_logs_first false   # When true fetch the oldest logs first
  drop_blank_events true    # Fluentd may throw an exception if a blank event is emitted
  telemetry false           # Produce statsd telemetry
  statsd_endpoint localhost # Endpoint to which telemetry should be sent
  <parse>
    @type cloudwatch_ingest
    expression /^(?<message>.+)$/
    time_format %Y-%m-%d %H:%M:%S.%L
    event_time true                             # take time from the Cloudwatch event, rather than parse it from the body
    inject_group_name true                      # inject the group name into the record
    inject_stream_name true                     # inject the stream name into the record
    inject_cloudwatch_ingestion_time field_name # inject the iso8601 `ingestion_time` as returned by the Cloudwatch API
    inject_plugin_ingestion_time field_name     # inject the iso8601 time at which the plugin ingested the event
    parse_json_body false                       # Attempt to parse the body as json and add structured fields from the result
    fail_on_unparsable_json false               # If the body cannot be parsed as json do not ingest the record
    telemetry false                             # Produce statsd telemetry
    statsd_endpoint localhost                   # Endpoint to which telemetry should be sent
  </parse>
</source>
```

### Authentication
The plugin will assume an IAM instance role. Without either of the `sts_*` options that role will be used for authentication. With those set the plugin will
attempt to `sts:AssumeRole` the `sts_arn`. This is useful for fetching logs from many accounts where the fluentd infrastructure lives in one single account.

### Prefixes
Both the `log_group_name_prefix` and `log_stream_name_prefix` may be omitted, in which case all groups and streams will be ingested. For performance reasons it is often desirable to set the `log_stream_name_prefix` to be today's date, managed by a configuration management system.

### State file
The state file is a YAML serialization of the current ingestion state. When running in a HA configuration this should be placed on a shared filesystem, such as EFS.
The state file is opened with an exclusive write call and as such also functions as a lock file in HA configurations. See below.

### HA Setup
When the state file is located on a shared filesystem an exclusive write lock will attempted each `interval`.
As such it is safe to run multiple instances of this plugin consuming from the same CloudWatch logging source without fear of duplication, as long as they share a state file.
In a properly configured auto-scaling group this provides for uninterrupted log ingestion in the event of a failure of any single node.

### JSON parsing
With the `parse_json_body` option set to `true` the plugin will attempt to parse the body of the log entry as JSON. If this is successful any field/value pairs found will be added to the emitted record as structured fields.

If `fail_on_unparsable_json` is set to `true` a record body consisting of malformed json will cause the record to be rejected. You may wish to leave this setting as false if the plugin is ingesting multiple log groups with a mixture of json/structured and unstructured content.

The `expression` is applied before JSON parsing is attempted. One may therefore extract a JSON fragment from within the event body if it is decorated with additional free-form text.

### High volume Log Groups
If you're having ingestion problems from high volume log groups you're advised to enable telemetry in both the main plugin and the parser, and to also set both `inject_cloudwatch_ingestion_time` and `inject_plugin_ingestion_time` to `true`.

This will enable your telemetry system to plot the state of your rate limiting, the effect of the ingestion delay _inside_ Cloudwatch Logs (`timestamp` vs `ingestion_time`) and take appropriate tuning action.

### Telemetry
With `telemetry` set to `true` and a valid `statsd_endpoint` the plugin will emit telemetry in statsd format to 8125:UDP. It is up to you to configure your statsd-speaking daemon to add any prefix or tagging that you might want.

The metrics emitted in this version are:

```
api.calls.describeloggroups.attempted
api.calls.describeloggroups.failed
api.calls.describelogstreams.attempted
api.calls.describelogstreams.failed
api.calls.describeloggroups.excluded # due to log_group_exclude_regexp
api.calls.getlogevents.attempted
api.calls.getlogevents.failed
api.calls.getlogevents.invalid_token
events.emitted.success
events.emitted.blocked
```

Likewise when telemetry is enabled for the parser, the emitted metrics are:

```
parser.record.attempted
parser.record.success
parser.json.success     # if json parsing is enabled
parser.json.failed      # if json parsing is enabled
parser.ingestion_skew   # the difference between `timestamp` and `ingestion_time` as returned by the Cloudwatch API
parser.plugin_skew      # the difference between "now" and `timestamp`
```

### Sub-second timestamps
When using `event_time true` the `@timestamp` field for the record is taken from the time recorded against the event by Cloudwatch. This is the most common mode to run in as it's an easy path to normalization: all of your Lambdas or other AWS service need not have the same, valid, `time_format` nor a regex that matches every case.

If your output plugin supports sub-second precision (and you're running fluentd 0.14.x) you'll "enjoy" sub-second precision.

#### Elasticsearch
It is a common pattern to use fluentd alongside the [fluentd-plugin-elasticsearch](https://github.com/uken/fluent-plugin-elasticsearch) plugin, either directly or via [fluent-plugin-aws-elasticsearch-service](https://github.com/atomita/fluent-plugin-aws-elasticsearch-service), to ingest logs into Elasticsearch.

Prior to version 1.9.5 there was a bug within that plugin which, via an unwise cast, caused records without a named timestamp field to be cast to `DateTime`, losing the precision. This PR: https://github.com/uken/fluent-plugin-elasticsearch/pull/249 fixed that issue.

### IAM
IAM is a tricky and often bespoke subject. Here's a starter that will ingest all of the logs for all of your Lambdas in the account in which the plugin is running:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "logs:DescribeLogGroups",
                "logs:DescribeLogStreams",
                "logs:DescribeMetricFilters",
                "logs:FilterLogEvents",
                "logs:GetLogEvents"
            ],
            "Resource": [
                "arn:aws:logs:eu-west-1:123456789012:log-group:/aws/lambda/*:*"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "logs:DescribeLogGroups",
            ],
            "Resource": [
                "arn:aws:logs:eu-west-1:123456789012:log-group:*:*"
            ]
        }
    ]
}
```

### Cross-account authentication
Broadly speaking the IAM instance role of the host on which the plugin is running needs to be able to `sts:AssumeRole` the `sts_arn` (and obviously needs `sts_enabled` to be true).

The assumed role should look more-or-less like that above in terms of the actions and resource combinations required.

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/sampointer/fluent-plugin-cloudwatch-ingest.



CHANGELOG.md
# Changelog

## 0.1.3

* Initial release

## 0.2.1

* AWS SDK logging
* Code reorganization

## 0.3.1

* Limit events per API call
* Parser constructor fix (@snltd)

## 0.4.0

* Optionally fetch oldest logs first (@chaeyk)

## 0.5.4

* Optionally parse the body as JSON into structured fields

## 0.6.0

* Add statsd telemetry

## 1.0.0

* Print a stack trace when recusing exceptions (@chaeyk)
* If stored API token is invalid or corrupt, use a stored timestamp (@chaeyk)
* Truncate statefile before saving (@chaeyk)
* Amend how `api_interval` is used (see README.md) (@chaeyk)
* Improve null stream detection (@chaeyk)
* Remove streams from state file that are no longer present (@chaeyk)
* Apply `error_interval` when failing to get statefile lock (@chaeyk)
* `api_interval` deprecated in favour of `error_interval`

## 1.1.0

* Update aws-sdk runtime dependency

## 1.2.0

* Add the ability to inject both the `ingestion_time returned from the the Cloudwatch Logs API, and the time that this plugin ingested the event into the record.
* Add telemetry to the parser

Both of these changes are designed to make debugging ingestion problems from high-volume log groups easier.

## 1.3.0

* Add `log_group_exclude_regexp` to allow optional exclusion of log groups by regexp

## 1.4.0

* Refuse to emit records with a blank (or newline only) message
* Emit metric `events.emitted.blocked` to expose these alongside logging
* Add plugin skew time to telemetry optionally emitted from the parser

## 1.5.0

* Limit the number of streams to be processed with each log stream describe call
* new parameter `max_log_streams_per_group` with a default of 50 (the default value for *limit* on API calls). This can be increased or decreased to limit the throttling of calls to AWS API
* Bail out of processing if fluentd has been stopped
* Move to the modular v3 of the aws-sdk

# 1.6.0
* Move back to v2 of the aws-sdk. The rest of the ecosystem just is not ready to move and I think pushing this here will create unnecessary pain.


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'fluent/plugin/cloudwatch/ingest'


spec/fluent/plugin/cloudwatch/ingest.rb
require 'fluent/plugin/cloudwatch/ingest/version'

module Fluent
  module Plugin
    module Cloudwatch
      module Ingest
        # Your code goes here...
      end
    end
  end
end


spec/fluent/plugin/cloudwatch/ingest_spec.rb
require 'spec_helper'

describe Fluent::Plugin::Cloudwatch::Ingest do
  it 'has a version number' do
    expect(Fluent::Plugin::Cloudwatch::Ingest::VERSION).not_to be nil
  end

  it 'does something useful' do
    expect(false).to eq(false)
  end
end


lib/fluent/plugin/parser_cloudwatch_ingest.rb
require 'date'
require 'fluent/plugin/parser_regexp'
require 'fluent/time'
require 'multi_json'
require 'statsd-ruby'

module Fluent
  module Plugin
    class CloudwatchIngestParser < RegexpParser
      Plugin.register_parser('cloudwatch_ingest', self)

      config_param :expression, :string, default: '^(?<message>.+)$'
      config_param :time_format, :string, default: '%Y-%m-%d %H:%M:%S.%L'
      config_param :event_time, :bool, default: true
      config_param :inject_group_name, :bool, default: true
      config_param :inject_cloudwatch_ingestion_time, :string, default: false
      config_param :inject_plugin_ingestion_time, :string, default: false
      config_param :inject_stream_name, :bool, default: true
      config_param :parse_json_body, :bool, default: false
      config_param :fail_on_unparsable_json, :bool, default: false
      config_param :telemetry, :bool, default: false
      config_param :statsd_endpoint, :string, default: 'localhost'

      def initialize
        super
      end

      def configure(conf)
        super
        @statsd = Statsd.new @statsd_endpoint, 8125 if @telemetry
      end

      def metric(method, name, value = 0)
        case method
        when :increment
          @statsd.send(method, name) if @telemetry
        else
          @statsd.send(method, name, value) if @telemetry
        end
      end

      def parse(event, group, stream)
        metric(:increment, 'parser.record.attempted')

        time = nil
        record = nil
        super(event.message) do |t, r|
          time = t
          record = r
        end

        # Optionally attempt to parse the body as json
        if @parse_json_body
          begin
            # Whilst we could just merge! the parsed
            # message into the record we'd bork on
            # nested keys. Force level one Strings.
            json_body = MultiJson.load(record['message'])
            metric(:increment, 'parser.json.success')
            json_body.each_pair do |k, v|
              record[k.to_s] = v.to_s
            end
          rescue MultiJson::ParseError
            metric(:increment, 'parser.json.failed')
            if @fail_on_unparsable_json
              yield nil, nil
              return
            end
          end
        end

        # Inject optional fields
        record['log_group_name'] = group if @inject_group_name
        record['log_stream_name'] = stream if @inject_stream_name

        if @inject_plugin_ingestion_time
          now = DateTime.now
          record[@inject_plugin_ingestion_time] = now.iso8601
        end

        if @inject_cloudwatch_ingestion_time
          epoch_ms = event.ingestion_time.to_f / 1000
          time = Time.at(epoch_ms)
          record[@inject_cloudwatch_ingestion_time] =
            time.to_datetime.iso8601(3)
        end

        # Optionally emit cloudwatch event and ingestion time skew telemetry
        if @telemetry
          metric(
            :gauge,
            'parser.ingestion_skew',
            event.ingestion_time - event.timestamp
          )
        end

        # Optionally emit @timestamp and plugin ingestion time skew
        if @telemetry
          metric(
            :gauge,
            'parser.plugin_skew',
            now.strftime('%Q').to_i - event.timestamp
          )
        end

        # We do String processing on the event time here to
        # avoid rounding errors introduced by floating point
        # arithmetic.
        event_s  = event.timestamp.to_s[0..9].to_i
        event_ns = event.timestamp.to_s[10..-1].to_i * 1_000_000

        time = Fluent::EventTime.new(event_s, event_ns) if @event_time

        metric(:increment, 'parser.record.success')
        yield time, record
      end
    end
  end
end


lib/fluent/plugin/in_cloudwatch_ingest.rb
# rubocop:disable Metrics/BlockLength
require 'aws-sdk-core'
require 'aws-sdk-cloudwatch'
require 'fluent/config/error'
require 'fluent/plugin/input'
require 'fluent/plugin/parser'
require 'json'
require 'pathname'
require 'psych'
require 'statsd-ruby'

module Fluent::Plugin
  class CloudwatchIngestInput < Fluent::Plugin::Input
    Fluent::Plugin.register_input('cloudwatch_ingest', self)
    helpers :compat_parameters, :parser

    desc 'The region of the source cloudwatch logs'
    config_param :region, :string, default: 'us-east-1'
    desc 'Enable STS for cross-account IAM'
    config_param :sts_enabled, :bool, default: false
    desc 'The IAM role ARN in the source account to use when STS is enabled'
    config_param :sts_arn, :string, default: ''
    desc 'The session name for use with STS'
    config_param :sts_session_name, :string, default: 'fluentd'
    desc 'Log group name or prefix. Not setting means "all"'
    config_param :log_group_name_prefix, :string, default: ''
    desc 'Log stream name or prefix. Not setting means "all"'
    config_param :log_stream_name_prefix, :string, default: ''
    desc 'Log group regexp to exclude, despite matching'
    config_param :log_group_exclude_regexp, :string, default: ''
    desc 'State file name'
    config_param :state_file_name, :string, default: \
      '/var/spool/td-agent/cloudwatch.state'
    desc 'Fetch logs every interval'
    config_param :interval, :time, default: 60
    desc 'Time to pause between error conditions'
    config_param :error_interval, :time, default: 5
    config_param :api_interval, :time
    desc 'Tag to apply to record'
    config_param :tag, :string, default: 'cloudwatch'
    desc 'Enable AWS SDK logging'
    config_param :aws_logging_enabled, :bool, default: false
    desc 'Limit the number of events fetched in any iteration'
    config_param :limit_events, :integer, default: 10_000
    desc 'Limit the number of log streams to be processed with each iteration'
    config_param :max_log_streams_per_group, :integer, default: 50
    desc 'Do not fetch events before this time'
    config_param :event_start_time, :integer, default: 0
    desc 'Fetch the oldest logs first'
    config_param :oldest_logs_first, :bool, default: false
    desc 'Refuse to emit events with a blank "message" field'
    config_param :drop_blank_events, :bool, default: true
    desc 'Turn on telemetry'
    config_param :telemetry, :bool, default: false
    desc 'Statsd endpoint to which telemetry should be written'
    config_param :statsd_endpoint, :string, default: 'localhost'
    config_section :parse do
      config_set_default :@type, 'cloudwatch_ingest'
      desc 'Regular expression with which to parse the event message'
      config_param :expression, :string, default: '^(?<message>.+)$'
      desc 'Take the timestamp from the event rather than the expression'
      config_param :event_time, :bool, default: true
      desc 'Time format to use when parsing event message'
      config_param :time_format, :string, default: '%Y-%m-%d %H:%M:%S.%L'
      desc 'Inject the log group name into the record'
      config_param :inject_group_name, :bool, default: true
      desc 'Inject the log stream name into the record'
      config_param :inject_stream_name, :bool, default: true
    end

    def initialize
      super
      log.info('Starting fluentd-plugin-cloudwatch-ingest')
    end

    def metric(method, name, value = 0)
      case method
      when :increment
        @statsd.send(method, name) if @telemetry
      else
        @statsd.send(method, name, value) if @telemetry
      end
    end

    def configure(conf)
      super
      compat_parameters_convert(conf, :parser)
      parser_config = conf.elements('parse').first
      unless parser_config
        raise Fluent::ConfigError, '<parse> section is required.'
      end
      unless parser_config['expression']
        raise Fluent::ConfigError, 'parse/expression is required.'
      end
      unless parser_config['event_time']
        raise Fluent::ConfigError, 'parse/event_time is required.'
      end

      # Configure telemetry, if enabled
      @statsd = Statsd.new @statsd_endpoint, 8125 if @telemetry

      # Fixup deprecated options
      if @api_interval
        @error_interval = @api_interval
        log.warn('api_interval is deprecated for error_interval')
      end

      @parser = parser_create(conf: parser_config)
      log.info('Configured fluentd-plugin-cloudwatch-ingest')

      @log_streams_next_token = nil
    end

    def start
      super
      log.info('Started fluentd-plugin-cloudwatch-ingest')

      # Get a handle to Cloudwatch
      aws_options = {}
      Aws.config[:region] = @region
      Aws.config[:logger] = log if @aws_logging
      log.info("Working in region #{@region}")

      if @sts_enabled
        aws_options[:credentials] = Aws::AssumeRoleCredentials.new(
          role_arn: @sts_arn,
          role_session_name: @sts_session_name
        )

        log.info('Using STS for authentication with source account ARN: '\
                 "#{@sts_arn}, session name: #{@sts_session_name}")
      else
        log.info('Using local instance IAM role for authentication')
      end
      @aws = Aws::CloudWatchLogs::Client.new(aws_options)
      @finished = false
      @thread = Thread.new(&method(:run))
    end

    def shutdown
      @finished = true
      @thread.join
    end

    private

    def emit(event, log_group_name, log_stream_name)
      @parser.parse(event, log_group_name, log_stream_name) do |time, record|
        if record['message'].chomp.empty? && @drop_blank_events
          log.warn(
            'Event is blank or contains only a newline, refusing to '\
            "emit. group: #{log_group_name}, "\
            "stream: #{log_stream_name}, event: #{event}"\
          )
          metric(:increment, 'events.emitted.blocked')
          next
        end
        router.emit(@tag, time, record)
        metric(:increment, 'events.emitted.success')
      end
    end

    def log_groups(log_group_prefix)
      log_groups = []

      # Fetch all log group names
      next_token = nil
      loop do
        begin
          metric(:increment, 'api.calls.describeloggroups.attempted')
          response = if !log_group_prefix.empty?
                       @aws.describe_log_groups(
                         log_group_name_prefix: log_group_prefix,
                         next_token: next_token
                       )
                     else
                       @aws.describe_log_groups(
                         next_token: next_token
                       )
                     end

          regex = Regexp.new(@log_group_exclude_regexp)
          response.log_groups.each do |group|
            if !@log_group_exclude_regexp.empty?
              if regex.match(group.log_group_name)
                log.info(
                  "Excluding log_group #{group.log_group_name} due to "\
                  "log_group_exclude_regexp #{@log_group_exclude_regexp}"\
                )
                metric(:increment, 'api.calls.describeloggroups.excluded')
              else
                log_groups << group.log_group_name
              end
            else
              log_groups << group.log_group_name
            end
          end
          break unless response.next_token
          next_token = response.next_token
        rescue StandardError => boom
          log.error("Unable to retrieve log groups: #{boom.inspect}")
          metric(:increment, 'api.calls.describeloggroups.failed')
          next_token = nil
          sleep @error_interval
          retry
        end
      end
      log.info("Found #{log_groups.size} log groups")

      return log_groups
    end

    def log_streams(log_group_name, log_stream_name_prefix)
      log_streams = []

      begin
        metric(:increment, 'api.calls.describelogstreams.attempted')
        response = if !log_stream_name_prefix.empty?
                     @aws.describe_log_streams(
                       log_group_name: log_group_name,
                       log_stream_name_prefix: log_stream_name_prefix,
                       next_token: @log_streams_next_token,
                       limit: @max_log_streams_per_group
                     )
                   else
                     @aws.describe_log_streams(
                       log_group_name: log_group_name,
                       next_token: @log_streams_next_token,
                       limit: @max_log_streams_per_group
                     )
                   end

        response.log_streams.each { |s| log_streams << s.log_stream_name }
        if log_streams.size == @max_log_streams_per_group
          @log_streams_next_token = response.next_token
        end
      rescue StandardError => boom
        prefix_message = if log_stream_name_prefix.empty?
                           ''
                         else
                           "with stream prefix #{log_stream_name_prefix}"
                         end

        log.error('Unable to retrieve log streams '\
                  "for group #{log_group_name} #{prefix_message}: "\
                  "#{boom.inspect}")
        metric(:increment, 'api.calls.describelogstreams.failed')

        sleep @error_interval
        retry
      end

      log.info("Found #{log_streams.size} streams for #{log_group_name}")

      return log_streams
    end

    def process_stream(group, stream, next_token, start_time, state)
      event_count = 0

      metric(:increment, 'api.calls.getlogevents.attempted')
      response = @aws.get_log_events(
        log_group_name: group,
        log_stream_name: stream,
        next_token: next_token,
        limit: @limit_events,
        start_time: start_time,
        start_from_head: @oldest_logs_first
      )

      response.events.each do |e|
        begin
          emit(e, group, stream)
          event_count += 1
        rescue StandardError => boom
          log.error("Failed to emit event #{e}: #{boom.inspect}")
        end
      end

      log.info("#{event_count} events processed for stream #{stream}")

      has_stream_timestamp = true if state.store[group][stream]['timestamp']

      if !has_stream_timestamp && response.events.count.zero?
        # This stream has returned no data ever.
        # In this case, don't save state (token could be an invalid one)
      else
        # Once all events for this stream have been processed,
        # in this iteration, store the forward token
        state.new_store[group][stream]['token'] = response.next_forward_token
        if response.events.last
          state.new_store[group][stream]['timestamp'] =
            response.events.last.timestamp
        else
          state.new_store[group][stream]['timestamp'] =
            state.store[group][stream]['timestamp']
        end
      end

      return event_count
    end

    def run
      until @finished
        begin
          state = State.new(@state_file_name, log)
        rescue StandardError => boom
          log.info("Failed lock state. Sleeping for #{@error_interval}: "\
                   "#{boom.inspect}")
          sleep @error_interval
          next
        end

        event_count = 0
        # For each log group
        log_groups(@log_group_name_prefix).each do |group|
          loop do
            begin
              # Fetch log group streams and emit the events
              log_streams(group, @log_stream_name_prefix).each do |stream|
                unless state.store[group][stream]
                  state.store[group][stream] = {}
                end

                log.info("processing stream: #{stream}")

                # See if we have some stored state for this group and stream.
                # If we have then use the stored forward_token to pick up
                # from that point. Otherwise start from the start.

                begin
                  event_count += process_stream(
                    group,
                    stream,
                    state.store[group][stream]['token'],
                    @event_start_time,
                    state
                  )
                rescue Aws::CloudWatchLogs::Errors::InvalidParameterException
                  metric(:increment, 'api.calls.getlogevents.invalid_token')
                  log.error(
                    'cloudwatch token is expired or broken. '\
                    'trying with timestamp.'
                  )

                  # try again with timestamp instead of forward token
                  begin
                    timestamp = state.store[group][stream]['timestamp']
                    timestamp ||= @event_start_time

                    event_count += process_stream(group,
                                                  stream,
                                                  nil,
                                                  timestamp,
                                                  state)
                  rescue StandardError => boom
                    log.error(
                      'Unable to retrieve events for stream '\
                      "#{stream} in group #{group}: "\
                      "#{boom.inspect}"\
                    )
                    metric(:increment, 'api.calls.getlogevents.failed')
                    sleep @error_interval
                    next
                  end
                rescue StandardError => boom
                  log.error("Unable to retrieve events for stream #{stream} "\
                            "in group #{group}: #{boom.inspect}")
                  metric(:increment, 'api.calls.getlogevents.failed')
                  sleep @error_interval
                  next
                end
              end
              # process log streams in batches
              break if @log_streams_next_token.nil? || @finished
            end
            log.info("#{event_count} events processed for group #{group}")
          end
        end

        log.info('Saving state')
        begin
          state.save
          state.close
        rescue StandardError => boom
          log.error("Unable to save state file: #{boom.inspect}")
        end

        if !@finished # no need to sleep if it's finished
          sleep_interval = if event_count > 0
                             @interval
                           else
                             @error_interval # slow down when no events
                           end

          log.info("Pausing for #{sleep_interval}")
          sleep sleep_interval
        else
          log.info('Stopping fluentd-plugin-cloudwatch-ingest')
        end
      end
    end

    class CloudwatchIngestInput::State
      class LockFailed < RuntimeError; end
      attr_accessor :statefile, :store, :new_store

      def initialize(filepath, log)
        @filepath = filepath
        @log = log
        @store = Hash.new { |h, k| h[k] = Hash.new { |x, y| x[y] = {} } }
        @new_store = Hash.new { |h, k| h[k] = Hash.new { |x, y| x[y] = {} } }

        if File.exist?(filepath)
          self.statefile = Pathname.new(@filepath).open('r+')
        else
          @log.warn("No state file #{statefile} Creating a new one.")
          begin
            self.statefile = Pathname.new(@filepath).open('w+')
            save
          rescue StandardError => boom
            @log.error("Unable to create new file #{statefile.path}: "\
                       "#{boom.inspect}")
          end
        end

        # Attempt to obtain an exclusive flock on the file and raise and
        # exception if we can't
        @log.info("Obtaining exclusive lock on state file #{statefile.path}")
        lockstatus = statefile.flock(File::LOCK_EX | File::LOCK_NB)
        raise CloudwatchIngestInput::State::LockFailed unless lockstatus

        begin
          @store.merge!(Psych.safe_load(statefile.read))

          # Migrate old state file
          @store.each_value do |streams|
            streams.update(streams) do |_name, stream|
              if stream.is_a? String
                return { 'token' => stream, 'timestamp' => Time.now.to_i }
              end
              return stream
            end
          end

          @log.info("Loaded #{@store.keys.size} groups from #{statefile.path}")
        rescue StandardError
          statefile.close
          raise
        end
      end

      def save
        statefile.rewind
        statefile.truncate(0)
        statefile.write(Psych.dump(@new_store))
        @log.info("Saved state to #{statefile.path}")
        statefile.rewind
      end

      def close
        statefile.close
      end
    end
  end
end
